[DRAFT] All Project Stein Changes So Far

anandagopal6 opened the pull request:
<!-- Please provide all the information below.  -->

### Issue describing the changes in this PR

resolves #issue_for_this_pr

### Pull request checklist

* [ ] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ ] I have added all required tests (Unit tests, E2E tests)

<!-- Optional: delete if not applicable  -->
### Additional information

Additional PR information

anandagopal6 committed:
Merge dev into branch containing all worker indexing code changes/additions.

anandagopal6 committed:
All changes

anandagopal6 committed:
Merge branch 't-anjanan/allChanges' of https://github.com/Azure/azure-functions-host into t-anjanan/allChanges

anandagopal6 committed:
Fix test failures by manually adding worker indexing environment variables/worker config variables to test cases.

Add null checks to _workerRuntime and workerConfig code sections.

anandagopal6 commented:
/azp run

anandagopal6 committed:
Fix test failures related to FunctionMetadataProviderFactory initialization.

soninaren commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -228,6 +228,7 @@ internal void WorkerInitResponse(GrpcEvent initEvent)
 
             _workerChannelLogger.LogDebug("Received WorkerInitResponse. Worker process initialized");
             _initMessage = initEvent.Message.WorkerInitResponse;
+            _initMessage.Capabilities.Add("WorkerIndexing", "true"); // manually add capability for now without touching worker. Assume it is true.
```
This should be removed

soninaren commented on src/WebJobs.Script.Grpc/Channel/WorkerCapabilities.cs:
```diff
@@ -0,0 +1,47 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
```
Remove this file

anandagopal6 committed:
Remove extra line of code in GrpcWorkerChannel and try something for integration tests

soninaren commented on src/WebJobs.Script/Host/FunctionMetadataManager.cs:
```diff
@@ -31,9 +32,10 @@ public class FunctionMetadataManager : IFunctionMetadataManager
         private ImmutableArray<FunctionMetadata> _functionMetadataArray;
         private Dictionary<string, ICollection<string>> _functionErrors = new Dictionary<string, ICollection<string>>();
         private ConcurrentDictionary<string, FunctionMetadata> _functionMetadataMap = new ConcurrentDictionary<string, FunctionMetadata>(StringComparer.OrdinalIgnoreCase);
+        private IFunctionMetadataProviderFactory _functionMetadataProviderFactory;
```
nit: Factory could be readonly

soninaren commented on src/WebJobs.Script/Host/FunctionMetadataProviderFactory.cs:
```diff
@@ -0,0 +1,69 @@
+        public void Create()
+        {
+            _workerFunctionMetadataProvider = new WorkerFunctionMetadataProvider(_applicationHostOptions, _loggerFactory.CreateLogger<WorkerFunctionMetadataProvider>(), _metricsLogger);
+
```
nit: empty line

soninaren commented on src/WebJobs.Script/Host/FunctionMetadataManager.cs:
```diff
@@ -76,11 +80,11 @@ public bool TryGetFunctionMetadata(string functionName, out FunctionMetadata fun
         /// <param name="includeCustomProviders">Include any metadata provided by IFunctionProvider when loading the metadata</param>
         /// <returns> An Immmutable array of FunctionMetadata.</returns>
-        public ImmutableArray<FunctionMetadata> GetFunctionMetadata(bool forceRefresh, bool applyAllowList = true, bool includeCustomProviders = true)
+        public ImmutableArray<FunctionMetadata> GetFunctionMetadata(bool forceRefresh, bool applyAllowList = true, bool includeCustomProviders = true, IFunctionInvocationDispatcher dispatcher = null, bool workerIndexing = false)
```
workerIndexing is unused

soninaren commented on src/WebJobs.Script/Host/FunctionMetadataManager.cs:
```diff
@@ -117,14 +121,18 @@ private void InitializeServices()
         /// Read all functions and populate function metadata.
         /// </summary>
-        internal ImmutableArray<FunctionMetadata> LoadFunctionMetadata(bool forceRefresh = false, bool includeCustomProviders = true)
+        internal ImmutableArray<FunctionMetadata> LoadFunctionMetadata(bool forceRefresh = false, bool includeCustomProviders = true, IFunctionInvocationDispatcher dispatcher = null, bool workerIndexing = false)
```
workerIndexing is unused

soninaren reviewed (changes requested):
Requesting nit changes and others that we discussed on the call

anandagopal6 committed:
Remove unnecessary workerIndexing variable from FunctionMetadataManager

anandagopal6 commented:
/azp run

anandagopal6 committed:
Remove experimental line

anandagopal6 committed:
Fix build issue by removing unnecessary variable

anandagopal6 committed:
Add experimental line to see if this fixes integration/non-integration tests

anandagopal6 committed:
Removed _channels from RpcFunctionInvocationDispatcher, fixed boolean logic in FunctionMetadataProviderFactory

anandagopal6 committed:
Renaming GRPC messages, fix validation and FinishInitialization

anandagopal6 committed:
Renamed GrpcWorkerChannel worker indexing functions to match GRPC message names

Fixed invocation troubles for worker indexed functions --> was just missing a few lines from ScriptHost

anandagopal6 committed:
Removed WorkerCapabilities service

anandagopal6 committed:
Reorganized dispatcher references: now located in WorkerFunctionMetadataProvider constructor instead of passing it through the GetFunctionMetadata() calls

anandagopal6 commented:
Closing this in favor of the larger PR with all Stein changes
